### PR TITLE
fix #247 make _Config class a dataclass and remote type:ignore

### LINKED ISSUE #247: Make _Config in runner.py class a dataclass so we can remove the whole lot of type: ignore-s from later parts of cli.py
This not being a dataclass causes a lots of confusion for mypy, and rightfully so. This is not the nicest. Make it a dataclass or replace everything with click :)

### PR DESCRIPTION
**Pull request by xiaoxianma**
**What:**

**Why:**

**How:**

**Risks:**

**Checklist**:

- [x] Added tests, if you've added code that should be tested
- [x] Updated the documentation, if you've changed APIs
- [x] Ensured the test suite passes
- [x] Made sure your code lints
- [x] Completed the Contributor License Agreement ("CLA")

### CODE REVIEW

**Comment by xiaoxianma** [comment]
@deathowl Trying to use Click, there are more places to be refactored than I thought. I ended up with making _Config dataclass.

**Comment by xiaoxianma** [comment]
@deathowl Hmm, it seems that mypy failures are not introduced by this commit. Do you have an idea? Thanks

**Review by deathowl — APPROVED** [review]

**Comment by deathowl** [comment]
@xiaoxianma  Thank you for your contribution :)

File: testslide/__init__.py [hunk 1]
```diff
@@ -421,7 +421,7 @@ def showwarning(
             else:
                 original_showwarning(message, category, filename, lineno, file, line)
 
-        warnings.showwarning = showwarning
+        warnings.showwarning = showwarning  # type: ignore
```
Comment (fornellas) [inline testslide/__init__.py:424]: I'm curious why this ignore was needed now (same for `logger_warning` below). Is this one of the cases we could / should add an exception at `mypy.ini` as `ignore_missing_imports = True`?

Comment (deathowl) [reply testslide/__init__.py:424]: Nono, it's ```testslide/__init__.py:441: error: Cannot assign to a method```

Comment (fornellas) [reply testslide/__init__.py:424]: Something's odd. Just tried this on master (8833026):

- Remove these 2 `type: ignore`.
- Pyenv with Python 3.8.5.
- `python -m venv venv`
- `. venv/bin/activate`
- `make install_build_deps`
- `make mypy`

and... no errors reported! Not sure what extra you have at your machine to detect this issue, but I'd love to have it on master, lol... I think it is time start using [tox](https://tox.readthedocs.io/en/latest/) and have a file with all dependency versions frozen, which we update every release.

**Pull request merged**